82. Remove Duplicates from Sorted List II.md

```diff
@@ -0,0 +1,139 @@
+1st
+Remove Duplicates from Sorted List とほぼ同じだが、重複するものはすべて取り除く必要がある。
+ポインタが head からスタートすると head の値が重複しているときに対応できないため、ダミーを用意してダミー.next に head をつなげる。
+そうなると、ノードを一つずつたどる用のポインターと、重複のないノードをつなぐ用の 2 つがいる？
+時間計算量:O(n)
+空間計算量:O(1)
+
+```python
+class Solution:
+    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
+        dummy = ListNode(0)
+        dummy.next = head
+
+        current_node = head
+        last_unique_node = dummy
+
+        while current_node:
+            if current_node.next and current_node.val == current_node.next.val:
+                while current_node.next and current_node.val == current_node.next.val:
```

Comment:
ここ実際はO(n)時間なのに二重ループのせいでぱっと見O(n^2)に見えるので、自分が実務で書くならO(n)だよって旨をコメントに書くと思いました

Reply:
コードを見ても計算量が分かりにくいという理由ならば、計算量はあまり書かないと思います。
というのも、計算量は極限でのふるまいなので、計算時間ではなくて計算量を知りたいことってあまりないんですよ。
あえていうと、計算量が悪いので特殊な入力ではまずい場合があることの注釈などです。

Chrome のソースコードで計算量がコメントに書かれているものは C++ は数万ファイルあって100箇所くらい、Python は2万ファイル弱あって10箇所くらいで0.1%の桁です。

https://source.chromium.org/search?q=%2F%2F.*%5C%20O%5C(%20filepath:.*%5C.cc$%20-filepath:third_party

https://source.chromium.org/search?q=%23.*%5C%20O%5C(%20filepath:.*%5C.py$%20-filepath:third_party

Reply:
ありがとうございます。
python側のコードを見ましたが、確かに計算量が言及されている箇所は
> Given that the BST strategy is O(n^3), don't choose a crazy large

> This is O(N*log(N)), although this can be done in O(N) with fancy datastructures.

など大きすぎる値への注意や、もっと効率よくする方法があるなどの文脈が多いですね。
勉強になりました。

計算量が記述されているのはどのような場合か、どれくらいの割合か、などを実際のソースコードを見て調べようと思わなかったので、このあたりも自然に調べられるようになりたいと思います。

Reply:
@oda 
ありがとうございます！実際のコードベースで調べてみようと思わなかったので、勉強になりました。まだまだ調べる癖が足りないですね

> 計算時間ではなくて計算量を知りたいことってあまりないんですよ

とありますが、反対に計算時間を記載する場合はありますか？自分でChomium Code Searchで"execution time"で検索してみたのですが、ヒットせずでした

Reply:
なんらかの疑問を抱くようなコードならば時間を書くでしょうが、そんなに多いわけではないです。
https://source.chromium.org/search?q=%5B0-9%5D%5C%20%5Bumn%5D%3Fs%5C%20%20filepath:.*%5C.cc$%20-filepath:third_party

> For a 16x16 icon on an Intel Core i5 this function takes approximately 0.5 ms to run.
https://source.chromium.org/chromium/chromium/src/+/main:ui/gfx/color_analysis.cc;l=463?q=%5B0-9%5D%5C%20%5Bumn%5D%3Fs%5C%20%20filepath:.*%5C.cc$%20-filepath:third_party

Reply:
> I ran some benchmarks comparing the performance of each on my Windows 7 workstation.
https://source.chromium.org/chromium/chromium/src/+/main:net/base/network_change_notifier_win.cc;l=153?q=%5B0-9%5D%5C%20%5Bumn%5D%3Fs%5C%20%20filepath:.*%5C.cc$%20-filepath:third_party

https://source.chromium.org/search?q=%5B0-9%5D%5C%20(%5Bumn%5Ds%5C%20%7Csec%7Cmin%7Chour)%20filepath:.*%5C.cc$%20-filepath:third_party